core/rust/qdbr/src/parquet/qdb_metadata.rs

```diff
@@ -0,0 +1,223 @@
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+#![allow(dead_code)]
+
+use crate::parquet::col_type::ColumnType;
+use crate::parquet::error::{fmt_err, ParquetError, ParquetErrorCause, ParquetResult};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+pub const QDB_META_KEY: &str = "questdb";
+
+/// A constant field that serializes always as the same value in JSON.
+/// On deserialization, it checks that the value is the same as the constant.
+#[derive(PartialEq)]
+struct U32Const<const N: u32>;
+
+impl<const N: u32> Serialize for U32Const<N> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        N.serialize(serializer)
+    }
+}
+
+impl<const N: u32> Debug for U32Const<N> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{N}")
+    }
+}
+
+impl<'de, const N: u32> Deserialize<'de> for U32Const<N> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let n = u32::deserialize(deserializer)?;
+        if n != N {
+            return Err(serde::de::Error::custom(format!(
+                "expected {}, got {}",
+                N, n
+            )));
+        }
+        Ok(U32Const)
+    }
+}
+
+/// A basic "version"-only metadata struct that's exclusively used to select
+/// which version of the metadata struct to deserialize.
+/// This contains just the `version` field that all the other versions
+/// of the metadata struct must also contain as a `U32Const<N>`.
+#[derive(Deserialize)]
+struct VersionMeta {
+    pub version: u32,
+}
+
+/// Special instructions on how to handle the column data,
+/// beyond the basic column type.
+#[derive(Debug, PartialEq, Serialize, Deserialize, Copy, Clone)]
+pub enum QdbMetaColHandling {
+    /// For dict-encoded columns, the row-range local dict key
+    /// is the same as the QuestDB's global dict key.
+    /// Used for symbol columns.
+    LocalKeyIsGlobal,
```

Comment:
I'm unsure if parsing JSON each time we access a Parquet partition is a good idea performance-wise. As for the `LocalKeyIsGlobal` flag, we could store QDB format version, e.g. in `FileMetaData.version` or `created_by`. I'm also not sure if we really need `column_type` in the metadata. We already have the table metadata plus in Parquet we have `field_id` to map table columns to Parquet ones, so we know all column types. So, maybe we don't need the JSON / `key_value_metadata` at all?

@eugenels WDYT?

Reply:
> I'm unsure if parsing JSON each time we access a Parquet partition is a good idea performance-wise.
I really wouldn't sweat parsing a few bytes of JSON when accessing a file which is going to be megabytes of data. Parsing JSON, even with `serde_json` (the most convenient approach) is still in the order of hundreds of MiB/s.

> so we know all column types
Not quite, we still need custom metadata for anything that serializes down to the same thing. Without it we'd never be able to differenciate between INT and IPV4 types.

> Then we can check the version and if it's v1, all symbol columns are encoded with local-to-global matching codes.
This is true for an early v1, but we know that we will want to accept externally generated parquet files too in the near future. In those cases it would be far easier to have a flag encoded in the metadata for this.

The logic is then quite easy:
```python
if local_key_is_global:
    do_special_key_only_logic()
else:
    do_lookups_and_return_extra_values_hashmap()
```

Far less confusing and convoluted than having to check the version, remember if the questdb symbol type was derived or not and from there determine the appropriate `local_key_is_global`.

> So, maybe we don't need the JSON / key_value_metadata at all?
Well.. as clarified earlier, we certainly need to track our original type somewhere.
It was previously done in `__qdb_type_{id}` keys, but that's a bit of an unusual approach.

For context, this is what the kv metadata for a parquet file generated from Python Pandas looks like:

```python
{b'ARROW:schema': b'//////ADAAAQAAAAAAAKAA4ABgAFAAgACgAAAAABBAAQAAAAAAAKAAwA'
                  b'AAAEAAgACgAAAOACAAAEAAAAAQAAAAwAAAAIAAwABAAIAAgAAAAIAAAA'
                  b'EAAAAAYAAABwYW5kYXMAAKsCAAB7ImluZGV4X2NvbHVtbnMiOiBbeyJr'
                  b'aW5kIjogInJhbmdlIiwgIm5hbWUiOiBudWxsLCAic3RhcnQiOiAwLCAi'
                  b'c3RvcCI6IDEwLCAic3RlcCI6IDF9XSwgImNvbHVtbl9pbmRleGVzIjog'
                  b'W3sibmFtZSI6IG51bGwsICJmaWVsZF9uYW1lIjogbnVsbCwgInBhbmRh'
                  b'c190eXBlIjogInVuaWNvZGUiLCAibnVtcHlfdHlwZSI6ICJvYmplY3Qi'
                  b'LCAibWV0YWRhdGEiOiB7ImVuY29kaW5nIjogIlVURi04In19XSwgImNv'
                  b'bHVtbnMiOiBbeyJuYW1lIjogImlkIiwgImZpZWxkX25hbWUiOiAiaWQi'
                  b'LCAicGFuZGFzX3R5cGUiOiAiaW50NjQiLCAibnVtcHlfdHlwZSI6ICJp'
                  b'bnQ2NCIsICJtZXRhZGF0YSI6IG51bGx9LCB7Im5hbWUiOiAidHMiLCAi'
                  b'ZmllbGRfbmFtZSI6ICJ0cyIsICJwYW5kYXNfdHlwZSI6ICJkYXRldGlt'
                  b'ZSIsICJudW1weV90eXBlIjogImRhdGV0aW1lNjRbbnNdIiwgIm1ldGFk'
                  b'YXRhIjogbnVsbH0sIHsibmFtZSI6ICJuYW1lIiwgImZpZWxkX25hbWUi'
                  b'OiAibmFtZSIsICJwYW5kYXNfdHlwZSI6ICJjYXRlZ29yaWNhbCIsICJu'
                  b'dW1weV90eXBlIjogImludDgiLCAibWV0YWRhdGEiOiB7Im51bV9jYXRl'
                  b'Z29yaWVzIjogOCwgIm9yZGVyZWQiOiBmYWxzZX19XSwgImNyZWF0b3Ii'
                  b'OiB7ImxpYnJhcnkiOiAicHlhcnJvdyIsICJ2ZXJzaW9uIjogIjE1LjAu'
                  b'MCJ9LCAicGFuZGFzX3ZlcnNpb24iOiAiMi4xLjIifQADAAAAqAAAAGQA'
                  b'AAAUAAAAEAAYAAgABgAHAAwAEAAUABAAAAAAAAEFFAAAADwAAAAgAAAA'
                  b'BAAAAAAAAAAEAAAAbmFtZQAAAAAIAAgAAAAEAAgAAAAEAAAAjP///wAA'
                  b'AAEIAAAABAAEAAQAAADQ////AAABChAAAAAcAAAABAAAAAAAAAACAAAA'
                  b'dHMAAAAABgAIAAYABgAAAAAAAwAQABQACAAGAAcADAAAABAAEAAAAAAA'
                  b'AQIQAAAAHAAAAAQAAAAAAAAAAgAAAGlkAAAIAAwACAAHAAgAAAAAAAAB'
                  b'QAAAAAAAAAA=',
 b'pandas': b'{"index_columns": [{"kind": "range", "name": null, "start": 0, "'
            b'stop": 10, "step": 1}], "column_indexes": [{"name": null, "field'
            b'_name": null, "pandas_type": "unicode", "numpy_type": "object", '
            b'"metadata": {"encoding": "UTF-8"}}], "columns": [{"name": "id", '
            b'"field_name": "id", "pandas_type": "int64", "numpy_type": "int64'
            b'", "metadata": null}, {"name": "ts", "field_name": "ts", "pandas'
            b'_type": "datetime", "numpy_type": "datetime64[ns]", "metadata": '
            b'null}, {"name": "name", "field_name": "name", "pandas_type": "ca'
            b'tegorical", "numpy_type": "int8", "metadata": {"num_categories":'
            b' 8, "ordered": false}}], "creator": {"library": "pyarrow", "vers'
            b'ion": "15.0.0"}, "pandas_version": "2.1.2"}'}
```

Notice the `"vendor": `"custom meta string"` approach used consistently between different libs.

The idea to introduce JSON here is also to follow this pattern.

Reply:
> I really wouldn't sweat parsing a few bytes of JSON when accessing a file which is going to be megabytes of data.

Decoding will happen multiple times, potentially even for the same file, and the accessed data may be quite small, especially with time interval filters. The less we do to read metadata, the better. We also have many allocations in the Rust code, while in Java, we're much more conservative.

When it comes to writes, we'll be writing metadata multiple times into a single time due to O3 - for wide tables writing JSON many times will add some overhead.

> Not quite, we still need custom metadata for anything that serializes down to the same thing. Without it we'd never be able to differenciate between INT and IPV4 types.

We have unique column ids in `field_id`, so we know if a column is INT or IPv4. QDB column types written into Parquet metadata are only used for assertions/extra validation, so we'll be perfectly fine without it.

> Well.. as clarified earlier, we certainly need to track our original type somewhere.

Yes, we already track that - in the table metadata, externally from Parquet files.

> This is true for an early v1, but we know that we will want to accept externally generated parquet files too in the near future.

Partitions will be only written by QDB, so we're in full control of them. 3rd-party SW will be able to read them, but not write them. External Parquet files are meant to be read with the `read_parquet()` SQL function.

Reply:
> I'm unsure if parsing JSON each time we access a Parquet partition is a good idea performance-wise. As for the `LocalKeyIsGlobal` flag, we could store QDB format version, e.g. in `FileMetaData.version` or `created_by`. Then we can check the version and if it's v1, all symbol columns are encoded with local-to-global matching codes. I'm also not sure if we really need `column_type` in the metadata. We already have the table metadata plus in Parquet we have `field_id` to map table columns to Parquet ones, so we know all column types. So, maybe we don't need the JSON / `key_value_metadata` at all?
> 
> @eugenels WDYT?

dunno, I don't expect a noticeable impact of JSON parsing on overall performance.
QDB native column seems to be useful for validation. say on a partition attachment.

> Decoding will happen multiple times, potentially even for the same file, and the accessed data may be quite small, especially with time interval filters. The less we do to read metadata, the better.

metadata can be read independently of data decoding and can be cached. there is a way to optimise

>When it comes to writes, we'll be writing metadata multiple times into a single time due to O3 - for wide tables writing JSON many times will add some overhead.

yes, once per O3 commit.

Reply:
Ok, I'm still not convinced that we need per-symbol column format, but if both of you think the JSON is not a big deal, let's keep it.

Example of what we currently have:
```json
{"version":1,"schema":{"columns":{"1":{"column_type":8},"2":{"column_type":12,"handling":"LocalKeyIsGlobal"},"0":{"column_type":5}}}}
```

Can we at least use an int enum instead of these `LocalKeyIsGlobal` strings and rename `column_type` to just `type` and `handling` to `format`?

```json
{"version":1,"schema":{"columns":[{"type":5},{"type":8},{"type":12,"format":1}]}}
```
would be more compact. Notice that the `columns` object is now an array - the current format with an array converted into a JSON object is weird.

Reply:
Still a few details to hash out over this. Probably best jump on a call over this.

Reply:
This got hashed out over a call w/ Andrei:
- Kept JSON, but slimmed it down.
- Extra bonus: No longer reliant on the `field_id` logic during queries.
- The `JSON` -> `QdbMeta` parsing will be extracted to a separate API over JNI